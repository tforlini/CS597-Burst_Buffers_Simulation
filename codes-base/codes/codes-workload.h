/*
 * Copyright (C) 2013 University of Chicago.
 * See COPYRIGHT notice in top-level directory.
 *
 */

/* I/O workload generator API to be used for reading I/O operations into
 * storage system simulations.  This API just describes the operations to be
 * executed; it does not service the operations.
 */

#ifndef CODES_WORKLOAD_H
#define CODES_WORKLOAD_H

#include "ross.h"
#define MAX_NAME_LENGTH_WKLD 512

typedef struct iolang_params iolang_params;
typedef struct darshan_params darshan_params;
typedef struct recorder_params recorder_params;
typedef struct codes_workload_info codes_workload_info;

/* struct to hold the actual data from a single MPI event*/
typedef struct scala_trace_params scala_trace_params;
typedef struct dumpi_trace_params dumpi_trace_params;


struct iolang_params
{
    /* the rank count is defined in the workload config file */
    int num_cns;
    /* flag - use path to find kernel files relative to the metafile */
    int use_relpath;
    char io_kernel_meta_path[MAX_NAME_LENGTH_WKLD];
    /* set by config in the metadata path */
    char io_kernel_path[MAX_NAME_LENGTH_WKLD];
};

struct darshan_params
{
    char log_file_path[MAX_NAME_LENGTH_WKLD];
    int64_t aggregator_cnt;
};

struct recorder_params
{
    char trace_dir_path[MAX_NAME_LENGTH_WKLD];
    int64_t nprocs;
};


struct codes_workload_info
{
    int group_id; /* group id */
    int min_rank; /* minimum rank in the collective operation */
    int max_rank; /* maximum rank in the collective operation */
    int local_rank; /* local rank? never being used in the bg/p model */
    int num_lrank; /* number of ranks participating in the collective operation*/
};

struct scala_trace_params {
   char offset_file_name[MAX_NAME_LENGTH_WKLD];
   char nw_wrkld_file_name[MAX_NAME_LENGTH_WKLD];
};

struct dumpi_trace_params {
   char file_name[MAX_NAME_LENGTH_WKLD];
   int num_net_traces;
};


/* supported I/O operations */
enum codes_workload_op_type
{
    /* terminator; there are no more operations for this rank */
    CODES_WK_END = 1,
    /* sleep/delay to simulate computation or other activity */
    CODES_WK_DELAY,
    /* block until specified ranks have reached the same point */
    CODES_WK_BARRIER,

    /* IO operations */
    /* open */
    CODES_WK_OPEN,
    /* close */
    CODES_WK_CLOSE,
    /* write */
    CODES_WK_WRITE,
    /* read */
    CODES_WK_READ,

    /* network operations (modelled after MPI operations) */
    /* blocking send operation */
    CODES_WK_SEND,
    /* blocking recv operation */
    CODES_WK_RECV,
    /* non-blocking send operation */
    CODES_WK_ISEND,
    /* non-blocking receive operation */
    CODES_WK_IRECV,
    /* broadcast operation */
    CODES_WK_BCAST,
    /* Allgather operation */
    CODES_WK_ALLGATHER,
    /* Allgatherv operation */
    CODES_WK_ALLGATHERV,
    /* Alltoall operation */
    CODES_WK_ALLTOALL,
    /* Alltoallv operation */
    CODES_WK_ALLTOALLV,
    /* Reduce operation */
    CODES_WK_REDUCE,
    /* Allreduce operation */
    CODES_WK_ALLREDUCE,
    /* Generic collective operation */
    CODES_WK_COL,
    /* Waitall operation */
    CODES_WK_WAITALL,
    /* Wait operation */
    CODES_WK_WAIT,
    /* Waitsome operation */
    CODES_WK_WAITSOME,
    /* Waitany operation */
    CODES_WK_WAITANY,
    /* Testall operation */
    CODES_WK_TESTALL,

    /* for workloads that have events not yet handled
     * (eg the workload language) */
    CODES_WK_IGNORE
};

/* I/O operation paramaters */
struct codes_workload_op
{
    /* TODO: do we need different "classes" of operations to differentiate
     * between different APIs?
     */

    /* what type of operation this is */
    enum codes_workload_op_type op_type;
    /* currently only used by network workloads */
    double start_time;
    double end_time;
    double sim_start_time;

    /* parameters for each operation type */
    union
    {
        struct {
            double seconds;
	    double nsecs;
        } delay;
        struct {
            int count;  /* num ranks in barrier, -1 means "all" */
            int root;   /* root rank */
        } barrier;
        struct {
            uint64_t file_id;      /* integer identifier for the file */
            int create_flag;  /* file must be created, not just opened */
        } open;
        struct {
            uint64_t file_id;  /* file to operate on */
            off_t offset; /* offset and size */
            size_t size;
        } write;
        struct {
            uint64_t file_id;  /* file to operate on */
            off_t offset; /* offset and size */
            size_t size;
        } read;
        struct {
            uint64_t file_id;  /* file to operate on */
        } close;
        struct {
            /* TODO: not sure why source rank is here */
            int source_rank;/* source rank of MPI send message */
            int dest_rank; /* dest rank of MPI send message */
            int num_bytes; /* number of bytes to be transferred over the network */
            int data_type; /* MPI data type to be matched with the recv */
            int count; /* number of elements to be received */
            int tag; /* tag of the message */
            int16_t req_id;
        } send;
        struct {
            /* TODO: not sure why source rank is here */
            int source_rank;/* source rank of MPI recv message */
            int dest_rank;/* dest rank of MPI recv message */
            int num_bytes; /* number of bytes to be transferred over the network */
            int data_type; /* MPI data type to be matched with the send */
            int count; /* number of elements to be sent */
            int tag; /* tag of the message */
            int16_t req_id;
        } recv;
        /* TODO: non-stub for other collectives */
        struct {
            int num_bytes;
        } collective;
	struct {
	    int count;
	    int16_t* req_ids;
	} waits;
	struct {
	   int16_t req_id;
	} wait;
    }u;
};

/* load and initialize workload of of type "type" with parameters specified by
 * "params".  The rank is the caller's relative rank within the collection
 * of processes that will participate in this workload.
 *
 * This function is intended to be called by a compute node LP in a model
 * and may be called multiple times over the course of a
 * simulation in order to execute different application workloads.
 *
 * Returns and identifier that can be used to retrieve operations later.
 * Returns -1 on failure.
 */
int codes_workload_load(const char* type, const char* params, int rank);

/* Retrieves the next I/O operation to execute.  the wkld_id is the
 * identifier returned by the init() function.  The op argument is a pointer
 * to a structure to be filled in with I/O operation information.
 */
void codes_workload_get_next(int wkld_id, int rank, struct codes_workload_op *op);

/* Reverse of the above function. */
void codes_workload_get_next_rc(int wkld_id, int rank, const struct codes_workload_op *op);

/* Retrieve the number of ranks contained in a workload */
int codes_workload_get_rank_cnt(const char* type, const char* params);

/* for debugging/logging: print an individual operation to the specified file */
void codes_workload_print_op(FILE *f, struct codes_workload_op *op, int rank);

/* NOTE: there is deliberately no finalize function; we don't have any
 * reliable way to tell when a workload is truly done and will not
 * participate in further reverse computation.   The underlying generators
 * will shut down automatically once they have issued their last event.
 */

#endif /* CODES_WORKLOAD_H */

/*
 * Local variables:
 *  c-indent-level: 4
 *  c-basic-offset: 4
 * End:
 *
 * vim: ft=c ts=8 sts=4 sw=4 expandtab
 */
