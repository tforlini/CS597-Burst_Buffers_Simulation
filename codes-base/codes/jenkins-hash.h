/*
 * Copyright (C) 2013 University of Chicago.
 * See COPYRIGHT notice in top-level directory.
 *
 */

#ifndef __JENKINS_HASH__
#define __JENKINS_HASH__

#include <sys/types.h>
#include <stdint.h>

/* This header declares and exports Bob Jenkins hash functions define in lookup3.c.
 * lookup3.c was downloaded from:  http://www.burtleburtle.net/bob/c/lookup3.c
 * See: http://www.burtleburtle.net/bob/hash/doobs.html for more information.
 *
 * Export other functions in that library as necessary.
 */

void bj_hashlittle2(
  const void *key,       /* the key to hash */
  size_t      length,    /* length of the key */
  uint32_t   *pc,        /* IN: primary initval, OUT: primary hash */
  uint32_t   *pb);        /* IN: secondary initval, OUT: secondary hash */

/* bj_hashsize(shift) gives a hash table size that is a power of 2, good for bj_hashlittle2.
 * shift values are:
 *
 * 1024 -> 10
 * 4096 -> 12
 * 16384 -> 14
 * 65536 -> 16
 * 262144 -> 18
 * 1048576 -> 20
 */
#define bj_hashsize(n) ((uint32_t)1<<(n))

/* bj_hashmask(n) gives a mask reasonable for returning a hash between 0 and the hashtable size.
 */
#define bj_hashmask(n) (bj_hashsize(n)-1)

#endif

/*
 * Local variables:
 *  c-indent-level: 4
 *  c-basic-offset: 4
 * End:
 *
 * vim: ts=8 sts=4 sw=4 expandtab
 */
