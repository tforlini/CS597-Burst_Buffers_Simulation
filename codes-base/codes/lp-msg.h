/*
 * Copyright (C) 2014 University of Chicago.
 * See COPYRIGHT notice in top-level directory.
 *
*/

#ifndef LP_MSG_H
#define LP_MSG_H

#include "ross.h"

/* It is good practice to always include the src LPID, a unique message
 * identifier, and a "magic" number to ensure that the LP type receiving the
 * message is the intended recipient. This is formalized in the following
 * struct and is used in a few places where LP-to-LP communication is
 * abstracted */

typedef struct msg_header_s {
    tw_lpid src;
    int event_type;
    int magic; 
} msg_header;

/* data structure utilities */
void msg_set_header(int magic, int event_type, tw_lpid src, msg_header *h);

#endif /* end of include guard: LP_MSG_H */

/*
 * Local variables:
 *  c-indent-level: 4
 *  c-basic-offset: 4
 * End:
 *
 * vim: ts=8 sts=4 sw=4 expandtab
 */
