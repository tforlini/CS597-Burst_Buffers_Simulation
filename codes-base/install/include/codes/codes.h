/*
 * Copyright (C) 2013 University of Chicago.
 * See COPYRIGHT notice in top-level directory.
 *
 */


#ifndef CODES_H
#define CODES_H

#include <ross.h>
#include <assert.h>

#if 0
#define codes_event_new tw_event_new
#else
static inline tw_event * codes_event_new(
    tw_lpid dest_gid, 
    tw_stime offset_ts, 
    tw_lp * sender)
{
    tw_stime abs_ts = offset_ts + tw_now(sender); 
    assert(abs_ts < g_tw_ts_end);
    //printf("codes_event_new() abs_ts: %.9f\n", abs_ts);
    return(tw_event_new(dest_gid, offset_ts, sender));
}
#endif

/* TODO: validate what value we should use here */
/* Modeled latency for communication between local software components and
 * communication between daemons and hardware devices.  Measured in
 * nanoseconds.
 */
#define CODES_MIN_LATENCY 0.5
#define CODES_MAX_LATENCY 1.0
#define CODES_LATENCY_RANGE \
    (CODES_MAX_LATENCY-CODES_MIN_LATENCY)
static inline tw_stime codes_local_latency(tw_lp *lp)
{
    int r = g_tw_nRNG_per_lp-1;
    tw_stime tmp;

    tmp = g_tw_lookahead + CODES_MIN_LATENCY +
        tw_rand_unif(&lp->rng[r]) * CODES_LATENCY_RANGE;

    if (g_tw_synchronization_protocol == CONSERVATIVE &&
            (tw_now(lp) + g_tw_lookahead) >= (tw_now(lp) + tmp))
        tw_error(TW_LOC,
                "codes_local_latency produced a precision loss "
                "sufficient to fail lookahead check (conservative mode) - "
                "increase CODES_MIN_LATENCY/CODES_MAX_LATENCY. "
                "Now:%0.5le, lookahead:%0.5le, return:%0.5le\n",
                tw_now(lp), g_tw_lookahead, tmp);

    return(tmp);
}

static inline void codes_local_latency_reverse(tw_lp *lp)
{
    int r = g_tw_nRNG_per_lp-1;
    tw_rand_reverse_unif(&lp->rng[r]);
    return;
}

#endif /* CODES_H */

/*
 * Local variables:
 *  c-indent-level: 4
 *  c-basic-offset: 4
 * End:
 *
 * vim: ft=c ts=8 sts=4 sw=4 expandtab
 */
