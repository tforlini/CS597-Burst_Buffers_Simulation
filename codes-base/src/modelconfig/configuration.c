/*
 * Copyright (C) 2013 University of Chicago.
 * See COPYRIGHT notice in top-level directory.
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <errno.h>
#include <libgen.h>
#include "codes/configuration.h"
#include <ross.h>

#include "configfile.h"
#include "txt_configfile.h"

/*
 * Global to hold configuration in memory
 */
ConfigHandle config;

/* Global to hold LP configuration */
config_lpgroups_t lpconf;

int configuration_load (const char *filepath,
                        MPI_Comm comm,
                        ConfigHandle *handle)
{
    MPI_File   fh;
    MPI_Status status;
    MPI_Offset txtsize;
    FILE      *f = NULL;
    char      *txtdata = NULL;
    char      *error = NULL;
    int        rc = 0;
    char      *tmp_path = NULL;

    rc = MPI_File_open(comm, (char*)filepath, MPI_MODE_RDONLY, MPI_INFO_NULL, &fh);
    if (rc != MPI_SUCCESS) goto finalize;

    rc = MPI_File_get_size(fh, &txtsize);
    if (rc != MPI_SUCCESS) goto finalize;

    txtdata = (char*) malloc(txtsize);
    assert(txtdata);

    rc = MPI_File_read_all(fh, txtdata, txtsize, MPI_BYTE, &status);
    if (rc != MPI_SUCCESS) goto finalize;

#ifdef __APPLE__
    f = fopen(filepath, "r");
#else
    f = fmemopen(txtdata, txtsize, "rb");
#endif
    if (!f) { rc = 1; goto finalize; }

    *handle = txtfile_openStream(f, &error);
    if (error) { rc = 1; goto finalize; }

    /* NOTE: posix version overwrites argument :(. */
    tmp_path = strdup(filepath);
    assert(tmp_path);
    (*handle)->config_dir = strdup(dirname(tmp_path));
    assert((*handle)->config_dir);

    rc = configuration_get_lpgroups(handle, "LPGROUPS", &lpconf);

finalize:
    if (fh != MPI_FILE_NULL) MPI_File_close(&fh);
    if (f) fclose(f);
    free(txtdata);
    free(tmp_path);
    if (error) {
        fprintf(stderr, "config error: %s\n", error);
        free(error);
    }

    return rc;
}

int configuration_get_value(ConfigHandle *handle,
                            const char *section_name,
                            const char *key_name,
                            const char *annotation,
                            char *value,
                            size_t len)
{
    SectionHandle section_handle;
    int           rc;
    // reading directly from the config, so need to inject the annotation
    // directly into the search string
    char key_name_tmp[CONFIGURATION_MAX_NAME];
    char *key_name_full;
    if (annotation==NULL){
        // sorry const type... we promise we won't change you
        key_name_full = (char*) key_name;
    }
    else{
        if (snprintf(key_name_tmp, CONFIGURATION_MAX_NAME, "%s@%s",
                    key_name, annotation) >= CONFIGURATION_MAX_NAME) {
            fprintf(stderr,
                    "config error: name@annotation pair too long: %s@%s\n",
                    key_name, annotation);
            return 1;
        }
        else
            key_name_full = key_name_tmp;
    }

    rc = cf_openSection(*handle, ROOT_SECTION, section_name, &section_handle);
    if (rc != 1) return 0;

    rc = cf_getKey(*handle, section_handle, key_name_full, value, len);
    (void) cf_closeSection(*handle, section_handle);

    return rc;
}

int configuration_get_value_relpath(
        ConfigHandle *handle,
        const char * section_name,
        const char * key_name,
        const char *annotation,
        char *value,
        size_t length){
    char *tmp = (char*) malloc(length);

    int w = configuration_get_value(handle, section_name, key_name, annotation, tmp,
            length);
    if (w <= 0)
        return w;

    /* concat the configuration value with the directory */
    w = snprintf(value, length, "%s/%s", (*handle)->config_dir, tmp);

    free(tmp);
    return w;
}

int configuration_get_multivalue(ConfigHandle *handle,
                                 const char *section_name,
                                 const char *key_name,
                                 const char *annotation,
                                 char ***values,
                                 size_t *len)
{
    SectionHandle section_handle;
    int           rc;
    // reading directly from the config, so need to inject the annotation
    // directly into the search string
    char key_name_tmp[CONFIGURATION_MAX_NAME];
    char *key_name_full;
    if (annotation==NULL){
        // sorry const type... we promise we won't change you
        key_name_full = (char*) key_name;
    }
    else{
        if (snprintf(key_name_tmp, CONFIGURATION_MAX_NAME, "%s@%s",
                    key_name, annotation) >= CONFIGURATION_MAX_NAME) {
            fprintf(stderr,
                    "config error: name@annotation pair too long: %s@%s\n",
                    key_name, annotation);
            return 1;
        }
        else
            key_name_full = key_name_tmp;
    }

    rc = cf_openSection(*handle, ROOT_SECTION, section_name, &section_handle);
    if (rc != 1) return rc;

    rc = cf_getMultiKey(*handle, section_handle, key_name_full, values, len);
    (void) cf_closeSection(*handle, section_handle);

    return rc;
}

int configuration_get_value_int (ConfigHandle *handle,
                                 const char *section_name,
                                 const char *key_name,
                                 const char *annotation,
                                 int *value)
{
    char valuestr[256];
    int rc = 1;
    int r;

    r = configuration_get_value(handle,
                                section_name,
                                key_name,
                                annotation,
                                valuestr,
                                sizeof(valuestr));
    if (r > 0)
    {
        *value = atoi(valuestr);
        rc = 0;
    }

    return rc;
}

int configuration_get_value_uint (ConfigHandle *handle,
                                  const char *section_name,
                                  const char *key_name,
                                  const char *annotation,
                                  unsigned int *value)
{
    char valuestr[256];
    int rc = 1;
    int r;

    r = configuration_get_value(handle,
                                section_name,
                                key_name,
                                annotation,
                                valuestr,
                                sizeof(valuestr));
    if (r > 0)
    {
        *value = (unsigned int) atoi(valuestr);
        rc = 0;
    }

    return rc;
}

int configuration_get_value_longint (ConfigHandle *handle,
                                     const char *section_name,
                                     const char *key_name,
                                     const char *annotation,
                                     long int *value)
{
    char valuestr[256];
    int rc = 1;
    int r;

    r = configuration_get_value(handle,
                                section_name,
                                key_name,
                                annotation,
                                valuestr,
                                sizeof(valuestr));
    if (r > 0)
    {
        errno = 0;
        *value = strtol(valuestr, NULL, 10);
        rc = errno;
    }

    return rc;
}

int configuration_get_value_double (ConfigHandle *handle,
                                    const char *section_name,
                                    const char *key_name,
                                    const char *annotation,
                                    double *value)
{
    char valuestr[256];
    int rc = 1;
    int r;

    r = configuration_get_value(handle,
                                section_name,
                                key_name,
                                annotation,
                                valuestr,
                                sizeof(valuestr));
    if (r > 0)
    {
        errno = 0;
        *value = strtod(valuestr, NULL);
        rc = errno;
    }

    return rc;
}

static void check_add_anno(
        const char *anno,
        config_anno_map_t *map){
    if (anno[0] == '\0'){
        map->has_unanno_lp = 1;
    }
    else{
        uint64_t a = 0;
        for (; a < map->num_annos; a++){
            if (strcmp(map->annotations[a], anno) == 0){
                map->num_anno_lps[a]++;
                break;
            }
        }
        if (a == map->num_annos){
            // we have a new anno!
            assert(a < CONFIGURATION_MAX_ANNOS);
            map->annotations[a] = strdup(anno);
            map->num_annos++;
            map->num_anno_lps[a] = 1;
        } // else anno was already there, do nothing
    }
}
static void check_add_lp_type_anno(
        const char *lp_name,
        const char *anno,
        config_lpgroups_t *lpgroups){
    uint64_t lpt_anno = 0;
    for (; lpt_anno < lpgroups->lpannos_count; lpt_anno++){
        config_anno_map_t *map = &lpgroups->lpannos[lpt_anno];
        if (strcmp(map->lp_name, lp_name) == 0){
            check_add_anno(anno, map);
            break;
        }
    }
    if (lpt_anno == lpgroups->lpannos_count){
        // we haven't seen this lp type before
        assert(lpt_anno < CONFIGURATION_MAX_TYPES);
        config_anno_map_t *map = &lpgroups->lpannos[lpt_anno];
        // initialize this annotation map
        strcpy(map->lp_name, lp_name);
        map->num_annos = 0;
        map->has_unanno_lp = 0;
        memset(map->num_anno_lps, 0, 
                CONFIGURATION_MAX_ANNOS*sizeof(*map->num_anno_lps));
        check_add_anno(anno, map);
        lpgroups->lpannos_count++;
    }
}

int configuration_get_lpgroups (ConfigHandle *handle,
                                const char *section_name,
                                config_lpgroups_t *lpgroups)
{
    SectionHandle sh;
    SectionHandle subsh;
    SectionEntry se[10];
    SectionEntry subse[10];
    size_t se_count = 10;
    size_t subse_count = 10;
    int i, j, lpt;
    char data[256];

    memset (lpgroups, 0, sizeof(*lpgroups));

    int ret = cf_openSection(*handle, ROOT_SECTION, section_name, &sh);
    if (ret == -1)
        return -1;
    cf_listSection(*handle, sh, se, &se_count); 

#define CHECKED_STRTOL(_val, _field, _data) \
    do{ \
        errno = 0; \
        long int _rd = strtol(_data, NULL, 10); \
        if (_rd <= 0 || errno) \
            tw_error(TW_LOC, "bad value (expected positive integer) for " \
                    "\"%s\": %s\n", _field, _data); \
        else \
            _val = _rd; \
    }while(0);

    for (i = 0; i < se_count; i++)
    {
        //printf("section: %s type: %d\n", se[i].name, se[i].type);
        if (se[i].type == SE_SECTION)
        {
            subse_count = 10;
            cf_openSection(*handle, sh, se[i].name, &subsh);
            cf_listSection(*handle, subsh, subse, &subse_count);
            strncpy(lpgroups->lpgroups[i].name, se[i].name,
                    CONFIGURATION_MAX_NAME);
            lpgroups->lpgroups[i].repetitions = 1;
            lpgroups->lpgroups_count++;
            for (j = 0, lpt = 0; j < subse_count; j++)
            {
                if (subse[j].type == SE_KEY)
                {
                   cf_getKey(*handle, subsh, subse[j].name, data, sizeof(data));
                   //printf("key: %s value: %s\n", subse[j].name, data);
                   if (strcmp("repetitions", subse[j].name) == 0)
                   {
                       CHECKED_STRTOL(lpgroups->lpgroups[i].repetitions,
                               "repetitions", data);
		       //printf("\n Repetitions: %ld ", lpgroups->lpgroups[i].repetitions);
                   }
                   else
                   {
                       size_t s = sizeof(lpgroups->lpgroups[i].lptypes[lpt].name);
                       char *nm   = lpgroups->lpgroups[i].lptypes[lpt].name;
                       char *anno = lpgroups->lpgroups[i].lptypes[lpt].anno;
                       // assume these are lptypes and counts
                       strncpy(nm, subse[j].name, s-1);
                       lpgroups->lpgroups[i].lptypes[lpt].name[s-1] = '\0';

                       char *c = strchr(nm, '@');
                       if (c) {
                           strcpy(anno, c+1);
                           *c = '\0';
                       }
                       else {
                           anno[0] = '\0';
                       }
                       // add to anno map
                       check_add_lp_type_anno(nm, anno, lpgroups);
                       CHECKED_STRTOL(lpgroups->lpgroups[i].lptypes[lpt].count,
                               nm, data);
                       lpgroups->lpgroups[i].lptypes_count++;
                       lpt++;
                   }
                }
            }
            cf_closeSection(*handle, subsh);
        }
    }

    cf_closeSection(*handle, sh);
    
    return 0;
}

/*
 * Helper function - get the position in the LP annotation list of the
 * given annotation. Used for configuration schemes where an array of
 * configuration values is generated based on the annotations in
 * config_anno_map_t
 * If anno is not found or a NULL anno is passed in,
 * -1 is returned */
int configuration_get_annotation_index(const char *              anno,
                                       const config_anno_map_t * anno_map){
    if (anno == NULL) return -1;
    for (uint64_t i = 0; i < anno_map->num_annos; i++){
        if (!strcmp(anno, anno_map->annotations[i])){
            return (int)i;
        }
    }
    return -1;
}

/*
 * Local variables:
 *  c-indent-level: 4
 *  c-basic-offset: 4
 * End:
 *
 * vim: ts=8 sts=4 sw=4 expandtab
 */
