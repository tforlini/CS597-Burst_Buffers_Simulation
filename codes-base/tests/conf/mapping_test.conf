LPGROUPS 
{
    GRP1
    {
        repetitions="2";
        a="1";
        b="2";
        a@foo="1";
        c="1";
    }
    GRP2
    {
        repetitions="3";
        c@bar="1";
        b@foo="1";
        c="2";
        a@foo="1";
    }
}

PARAMS
{
    message_size="256";
}
