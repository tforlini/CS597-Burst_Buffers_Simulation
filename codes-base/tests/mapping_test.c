/*
 * Copyright (C) 2014 University of Chicago.
 * See COPYRIGHT notice in top-level directory.
 *
 */

#include <ross.h>
#include "codes/configuration.h"
#include "codes/codes_mapping.h"
#include "codes/codes.h"

/* define three types of lps for mapping test */

typedef struct state_s{
    int id_global, id_by_group, id_by_anno, id_by_group_anno;
    char group_name[MAX_NAME_LENGTH], lp_name[MAX_NAME_LENGTH],
         anno[MAX_NAME_LENGTH];
} state;

static void init(state *ns, tw_lp *lp){
    int dummy;
    // only need the names
    codes_mapping_get_lp_info(lp->gid, ns->group_name, &dummy,
            ns->lp_name, &dummy, ns->anno, &dummy, &dummy);
    const char * anno = codes_mapping_get_annotation_by_lpid(lp->gid);
    // annotation check
    if (    (anno == NULL && ns->anno[0] != '\0') ||
            (anno != NULL && strcmp(anno, ns->anno) != 0)){
        fprintf(stderr, "LP %lu: annotations don't match: "
                "_get_lp_info:\"%s\", _get_anno_by_lpid:\"%s\"\n",
                lp->gid, ns->anno, anno);
    }

    ns->id_global = codes_mapping_get_lp_relative_id(lp->gid, 0, 0);
    ns->id_by_group = codes_mapping_get_lp_relative_id(lp->gid, 1, 0);
    ns->id_by_anno = codes_mapping_get_lp_relative_id(lp->gid, 0, 1);
    ns->id_by_group_anno = codes_mapping_get_lp_relative_id(lp->gid, 1, 1);

    // relative ID check - looking up the gid works correctly
    tw_lpid id_from_global_rel =
        codes_mapping_get_lpid_from_relative(ns->id_global, NULL,
                ns->lp_name, NULL, 0);
    tw_lpid id_from_group_rel =
        codes_mapping_get_lpid_from_relative(ns->id_by_group, ns->group_name,
                ns->lp_name, NULL, 0);
    tw_lpid id_from_anno_rel =
        codes_mapping_get_lpid_from_relative(ns->id_by_anno, NULL,
                ns->lp_name, anno, 1);
    tw_lpid id_from_group_anno_rel =
        codes_mapping_get_lpid_from_relative(ns->id_by_group_anno, ns->group_name,
                ns->lp_name, anno, 1);
    if (lp->gid != id_from_global_rel){
        fprintf(stderr, "LP %lu (%s): "
                "global relative id %d doesn't match (got %lu)\n",
                lp->gid, ns->lp_name, ns->id_global, id_from_global_rel);
    }
    if (lp->gid != id_from_group_rel){
        fprintf(stderr, "LP %lu (%s): "
                "group %s relative id %d doesn't match (got %lu)\n",
                lp->gid, ns->lp_name, ns->group_name, ns->id_by_group,
                id_from_group_rel);
    }
    if (lp->gid != id_from_anno_rel){
        fprintf(stderr, "LP %lu (%s): "
                "anno \"%s\" relative id %d doesn't match (got %lu)\n",
                lp->gid, ns->lp_name, ns->anno, ns->id_by_group,
                id_from_anno_rel);
    }
    if (lp->gid != id_from_group_anno_rel){
        fprintf(stderr, "LP %lu (%s): "
                "group %s anno \"%s\" relative id %d doesn't match (got %lu)\n",
                lp->gid, ns->lp_name, ns->group_name, ns->anno, ns->id_by_group,
                id_from_group_anno_rel);
    }


    // output-based check - print out IDs, compare against expected
    char tmp[128];
    if (ns->anno == NULL || ns->anno[0]=='\0')
        tmp[0] = '\0';
    else
        sprintf(tmp, "@%s", ns->anno);
    printf("TEST2 %2lu %2d %2d %2d %2d %s%s\n", lp->gid, ns->id_global,
            ns->id_by_group, ns->id_by_anno, ns->id_by_group_anno, ns->lp_name,
            tmp);
}

tw_lptype a_lp = {
    (init_f) init,
    (pre_run_f) NULL,
    (event_f) NULL,
    (revent_f) NULL,
    (final_f)  NULL,
    (map_f) codes_mapping,
    sizeof(state),
};
tw_lptype b_lp = {
    (init_f) init,
    (pre_run_f) NULL,
    (event_f) NULL,
    (revent_f) NULL,
    (final_f) NULL,
    (map_f) codes_mapping,
    sizeof(state),
};
tw_lptype c_lp = {
    (init_f) init,
    (pre_run_f) NULL,
    (event_f) NULL,
    (revent_f) NULL,
    (final_f) NULL,
    (map_f) codes_mapping,
    sizeof(state),
};

static char conf_file_name[128] = {'\0'};
static const tw_optdef app_opt [] =
{
	TWOPT_GROUP("codes-mapping test case" ),
    TWOPT_CHAR("codes-config", conf_file_name, "name of codes configuration file"),
	TWOPT_END()
};

int main(int argc, char *argv[])
{
    tw_opt_add(app_opt);
    tw_init(&argc, &argv);

    if (!conf_file_name[0]){
        fprintf(stderr, "Expected \"codes-config\" option, please see --help.\n");
        MPI_Finalize();
        return 1;
    }
    if (configuration_load(conf_file_name, MPI_COMM_WORLD, &config)){
        fprintf(stderr, "Error loading config file %s.\n", conf_file_name);
        MPI_Finalize();
        return 1;
    }

    lp_type_register("a", &a_lp);
    lp_type_register("b", &b_lp);
    lp_type_register("c", &c_lp);

    codes_mapping_setup();

    printf("# test 2 format:\n"
           "# <lp> rel id <global <group-wise> <anno.-wise> <both> <lp name>\n");
            
    const char * lps[]    = {"a", "b", "c"};
    const char * groups[] = {NULL, "GRP1", "GRP2"};
    const char * annos[]  = {NULL, "foo", "bar"};
    char lpnm[128];
    for (int l = 0; l < 3; l++){
        for (int g = 0; g < 3; g++){
            for (int a = 0; a < 3; a++){
                if (annos[a]==NULL)
                    sprintf(lpnm, "%s", lps[l]);
                else
                    sprintf(lpnm, "%s@%s", lps[l], annos[a]);
                printf("TEST1 %2d %6s %s\n",
                    codes_mapping_get_lp_count(groups[g], 0, lps[l],
                        annos[a], 0),
                    groups[g], lpnm);
            }
            printf("TEST1 %2d %6s %s ignore annos\n",
                codes_mapping_get_lp_count(groups[g], 0, lps[l], NULL, 1),
                groups[g], lps[l]);
        }
    }

    tw_run();
    tw_end();

    return 0;
}

/*
 * Local variables:
 *  c-indent-level: 4
 *  c-basic-offset: 4
 * End:
 *
 * vim: ft=c ts=8 sts=4 sw=4 expandtab
 */
