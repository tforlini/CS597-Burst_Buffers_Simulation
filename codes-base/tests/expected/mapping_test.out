TEST1  2 (null) a
TEST1  5 (null) a@foo
TEST1  0 (null) a@bar
TEST1  7 (null) a ignore annos
TEST1  2   GRP1 a
TEST1  2   GRP1 a@foo
TEST1  0   GRP1 a@bar
TEST1  4   GRP1 a ignore annos
TEST1  0   GRP2 a
TEST1  3   GRP2 a@foo
TEST1  0   GRP2 a@bar
TEST1  3   GRP2 a ignore annos
TEST1  4 (null) b
TEST1  3 (null) b@foo
TEST1  0 (null) b@bar
TEST1  7 (null) b ignore annos
TEST1  4   GRP1 b
TEST1  0   GRP1 b@foo
TEST1  0   GRP1 b@bar
TEST1  4   GRP1 b ignore annos
TEST1  0   GRP2 b
TEST1  3   GRP2 b@foo
TEST1  0   GRP2 b@bar
TEST1  3   GRP2 b ignore annos
TEST1  8 (null) c
TEST1  0 (null) c@foo
TEST1  3 (null) c@bar
TEST1 11 (null) c ignore annos
TEST1  2   GRP1 c
TEST1  0   GRP1 c@foo
TEST1  0   GRP1 c@bar
TEST1  2   GRP1 c ignore annos
TEST1  6   GRP2 c
TEST1  0   GRP2 c@foo
TEST1  3   GRP2 c@bar
TEST1  9   GRP2 c ignore annos
TEST2  0  0  0  0  0 a
TEST2  1  0  0  0  0 b
TEST2  2  1  1  1  1 b
TEST2  3  1  1  0  0 a@foo
TEST2  4  0  0  0  0 c
TEST2  5  2  2  1  1 a
TEST2  6  2  2  2  2 b
TEST2  7  3  3  3  3 b
TEST2  8  3  3  1  1 a@foo
TEST2  9  1  1  1  1 c
TEST2 10  2  0  0  0 c@bar
TEST2 11  4  0  0  0 b@foo
TEST2 12  3  1  2  0 c
TEST2 13  4  2  3  1 c
TEST2 14  4  0  2  0 a@foo
TEST2 15  5  3  1  1 c@bar
TEST2 16  5  1  1  1 b@foo
TEST2 17  6  4  4  2 c
TEST2 18  7  5  5  3 c
TEST2 19  5  1  3  1 a@foo
TEST2 20  8  6  2  2 c@bar
TEST2 21  6  2  2  2 b@foo
TEST2 22  9  7  6  4 c
TEST2 23 10  8  7  5 c
TEST2 24  6  2  4  2 a@foo
